### PR TITLE
Added room max capacity to limit room size

### PR DESCRIPTION
**Pull request by YingLiang2**
Rooms now have a limit of how many players can join a room.
* A notification appears on the bottom left indicating if they try to join a full room.
* An alert appears when a user tries to join a full room from a link 
     *  This solution can be further improved in the future, but I think this is best for now, thoughts @jonathanplacatka ?
     
image
image

### CODE REVIEW

**Comment by YingLiang2** [comment]
I have removed the notifications and added a simple text indicating the room is full. 

I feel like the box is pretty empty with text a button, but I don't think there is much to do atm mainly because this is for MVP, but maybe in the future we should add a graphic or something to enhance this box more? 

Also @jonathanplacatka thoughts on the button label?

image

**Comment by YingLiang2** [comment]
I also just added a simple button to the main page which when clicked redirects to the multiplayer page. 

This is obviously temporary as we will be working on the main page very soon. This is just for convenience so we can easily get to the multiplayer page. 

image

File: server/src/room.ts [hunk 1]
```diff
@@ -7,13 +7,15 @@ export default class Room {
...
     playersFinished: number;
+    maxCapacity: number;
     
     constructor(roomID: string) {
         this.roomID = roomID;
         this.#players = new Map()
         this.gameStarted = false;
         this.numWords = 0;
         this.playersFinished = 0;
+        this.maxCapacity = 1;
```
Comment (jonathanplacatka) [inline server/src/room.ts]: was this just 1 for testing?

Comment (YingLiang2) [reply server/src/room.ts]: Yes, once this is approved, I'll change it back to 4

File: server/src/server.ts [hunk 2]
```diff
@@ -54,14 +54,16 @@ export default class GameServer {
     
         if(roomToJoin.gameStarted) { 
             socket.emit('joinRoom', false);
-        } else {
+        } else if (Object.keys(roomToJoin.getPlayers()).length < roomToJoin.maxCapacity) {
```
Comment (jonathanplacatka) [inline server/src/server.ts]: we should move this into a small function in room (similar to isEmpty). then you can just call room.isFull()  in this check

File: web/package.json [hunk 3]
```diff
@@ -10,6 +10,7 @@
   },
   "dependencies": {
     "@mantine/core": "^7.11.2",
+    "@mantine/notifications": "^7.11.2",
```
Comment (jonathanplacatka) [inline web/package.json]: should be removed

File: web/src/app/components/Game.tsx [hunk 4]
```diff
@@ -48,10 +49,17 @@ export default function Game({roomID}: GameProps) {
...
 
-        socket.on('joinRoom', (success) => {
+        socket.on('joinRoom', (success, response) => {
             if(!success) {
-                alert("Game in progress");
-                socket.disconnect();
+                if (response === "roomFull") {
+                    // alert(`Room ${roomID} is currently full.`)
+                    setGameState(GameState.RoomFull);
+                    // router.push('multiplayer')
```
Comment (jonathanplacatka) [inline web/src/app/components/Game.tsx]: delete commented out code

File: web/src/app/multiplayer/page.tsx [hunk 5]
```diff
@@ -41,16 +41,31 @@ export default function Multiplayer() {
...
         );
     });
 
-    const joinRoom = (roomID : string) => {
-        router.push(roomID)
+    const joinRoom = (room: Room, roomID : string) => {
+        if (Object.keys(room.players).length < room.maxCapacity) {
+            router.push(roomID)
+        } else {
+            //This code is not used atm. May be used in the future.
```
Comment (jonathanplacatka) [inline web/src/app/multiplayer/page.tsx]: if its unused then delete it

File: web/src/app/components/Game.tsx [hunk 6]
```diff
@@ -134,6 +138,14 @@ export default function Game({roomID}: GameProps) {
             {gameState === GameState.Loading &&
                 <Loader className="mt-40" color="white" />
             }
+            {gameState === GameState.RoomFull && (
```
Comment (jonathanplacatka) [inline web/src/app/components/Game.tsx:142]: can we move this into it's own component? like Error.tsx or Message.tsx or something?

then we can pass a message string as a prop, and use the same component for the "Game in Progress" error

Comment (jonathanplacatka) [reply web/src/app/components/Game.tsx:142]: then in that case we can call the game state GameState.Error, and pass the message string as a prop depending on the result of the joinRoom event. just will need to make a state variable for the error message

**Review by jonathanplacatka — APPROVED** [review]

**Pull request merged**